File: basic_dsa_three_hundred_fifteen.java
/**
 * Basic DSA Problems - Two Sum II (Sorted Array)
 * Find two indices that sum to target
 */
public class basic_dsa_three_hundred_fifteen {

    // Problem: Sorted array and target. Return 1-indexed [i, j] where nums[i]+nums[j]=target.

    /**
     * Two pointers: if sum < target, move left++; if sum > target, move right--.
     * Time O(n), Space O(1)
     */
    public static int[] twoSum(int[] numbers, int target) {
        int i = 0, j = numbers.length - 1;
        while (i < j) {
            int s = numbers[i] + numbers[j];
            if (s == target) return new int[]{i + 1, j + 1};
            if (s < target) i++;
            else j--;
        }
        return new int[]{-1, -1};
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Two Sum II (Input Array Is Sorted)");
        System.out.println("=".repeat(60));

        System.out.println("[2,7,11,15], 9 -> " + java.util.Arrays.toString(twoSum(new int[]{2, 7, 11, 15}, 9)));
        System.out.println("[2,3,4], 6 -> " + java.util.Arrays.toString(twoSum(new int[]{2, 3, 4}, 6)));
        System.out.println("[-1,0], -1 -> " + java.util.Arrays.toString(twoSum(new int[]{-1, 0}, -1)));
    }
}

File: basic_dsa_three_hundred_fifteen.py
"""
Basic DSA - Count of Smaller Numbers After Self
Given an integer array nums, return an array counts where counts[i] is the number of
smaller elements to the right of nums[i].
"""


def count_smaller(nums):
    """
    Fenwick Tree over coordinate-compressed values.
    Time Complexity: O(n log n)
    Space Complexity: O(n)
    """
    if not nums:
        return []

    vals = sorted(set(nums))
    idx = {v: i + 1 for i, v in enumerate(vals)}  # 1-based for BIT
    size = len(vals)

    bit = [0] * (size + 1)

    def add(i, delta):
        while i <= size:
            bit[i] += delta
            i += i & -i

    def sum_(i):
        s = 0
        while i > 0:
            s += bit[i]
            i -= i & -i
        return s

    res = [0] * len(nums)
    for i in range(len(nums) - 1, -1, -1):
        r = idx[nums[i]]
        res[i] = sum_(r - 1)
        add(r, 1)
    return res


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Count of Smaller Numbers After Self")
    print("=" * 60)

    print(count_smaller([5, 2, 6, 1]))  # [2, 1, 1, 0]
    print(count_smaller([-1, -1]))  # [0, 0]
